From c4b1b4bc6c18fc0e226ebda23844dba9a83d2717 Mon Sep 17 00:00:00 2001
From: Maxim Orlov <orlovmg@gmail.com>
Date: Wed, 2 Jul 2025 17:44:32 +0300
Subject: [PATCH v67 07/12] Be tidy, zero dead tuples in
 heap_page_prune_execute

Although I can't imagine what it could break, we can be tidy and zero dead
tuples in heap_page_prune_execute.

Author: Evgeny Voropaev <evgeny.voropaev@tantorlabs.com>
Author: Maxim Orlov <orlovmg@gmail.com>

Discussion: https://www.postgresql.org/message-id/CACG%3Deza1343rteW-CjmaKYGGavcTmiAMrhXfQFVXHrmHKNPq2g%40mail.gmail.com
---
 src/backend/access/heap/pruneheap.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/src/backend/access/heap/pruneheap.c b/src/backend/access/heap/pruneheap.c
index 21ccf7ab17..ba947eecfc 100644
--- a/src/backend/access/heap/pruneheap.c
+++ b/src/backend/access/heap/pruneheap.c
@@ -1623,6 +1623,17 @@ heap_prune_record_unchanged_lp_redirect(PruneState *prstate, OffsetNumber offnum
 	prstate->processed[offnum] = true;
 }
 
+static inline void
+clear_removed_tuple(Page page, ItemId lp)
+{
+	if (ItemIdHasStorage(lp))
+	{
+		Assert(ItemIdIsNormal(lp));		/* valid only for heap pages */
+
+		memset(PageGetItem(page, lp), 0, MAXALIGN(ItemIdGetLength(lp)));
+	}
+}
+
 /*
  * Perform the actual page changes needed by heap_page_prune_and_freeze().
  *
@@ -1705,6 +1716,10 @@ heap_page_prune_execute(Buffer buffer, bool lp_truncate_only,
 		Assert(HeapTupleHeaderIsHeapOnly(htup));
 #endif
 
+		/* just to be tidy  */
+		if (!repairFragmentation)
+			clear_removed_tuple(page, fromlp);
+
 		ItemIdSetRedirect(fromlp, tooff);
 	}
 
@@ -1737,6 +1752,10 @@ heap_page_prune_execute(Buffer buffer, bool lp_truncate_only,
 		}
 #endif
 
+		/* just to be tidy  */
+		if (!repairFragmentation)
+			clear_removed_tuple(page, lp);
+
 		ItemIdSetDead(lp);
 	}
 
@@ -1776,6 +1795,10 @@ heap_page_prune_execute(Buffer buffer, bool lp_truncate_only,
 
 #endif
 
+		/* just to be tidy  */
+		if (!repairFragmentation)
+			clear_removed_tuple(page, lp);
+
 		ItemIdSetUnused(lp);
 	}
 
-- 
2.51.0

